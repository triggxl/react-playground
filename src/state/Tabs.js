import React from 'react';

class Tabs extends React.Component {
  // to fix TypeError: Cannot read property 'map' of undefined
  static defaultProps = {tabs: []};
  state = {
    currentTabIndex : 0
  };
  handleButtonClick = index => {
    //to modify state
    this.setState({currentTabIndex: index})
  }
  // syntax of classes, no const or let inside an object; using the this keyword is pertinent to classes
  renderButtons() {
    return this.props.tabs.map((tab, index) => (
      // add ()=> to execute on event and not on actual load of page
      <button key={index} onClick={()=> this.handleButtonClick(index)}>
        {tab.name}
      </button>
    ))
  }
  renderContent() {
    return (
      <div className='content'>
        {this.state.currentTabIndex.content};
      </div>
    )
  }
  render() {
    return (
      <div>
        {this.renderButtons()}
        {/* explain how line 34 reads */}
        {!!this.props.tabs.length && this.renderContent()}
      </div>
    )
  }
}

export default Tabs;