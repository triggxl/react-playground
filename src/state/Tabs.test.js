import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Tabs from './Tabs';

configure({ adapter: new Adapter() });

describe(`Tabs Component`, () => {
  const TabsProp = [
    { name: 'First tab',
      content: 'Lorem ipsum dolor sit amet consectetur adipisicing elit. Laboriosam exercitationem quos consectetur expedita consequatur. Fugit, sapiente aspernatur corporis velit, dolor eum reprehenderit provident ipsam, maiores incidunt repellat! Facilis, neque doloremque.' },
    { name: 'Second tab',
      content: 'Laboriosam exercitationem quos consectetur expedita consequatur. Fugit, sapiente aspernatur corporis velit, dolor eum reprehenderit provident ipsam, maiores incidunt repellat! Facilis, neque doloremque. Lorem ipsum dolor sit amet consectetur adipisicing elit.' },
    { name: 'Third tab',
      content: 'Fugit, sapiente aspernatur corporis velit, dolor eum reprehenderit provident ipsam, maiores incidunt repellat! Facilis, neque doloremque. Lorem ipsum dolor sit amet consectetur adipisicing elit. Laboriosam exercitationem quos consectetur expedita consequatur.' },
  ];
  //smoke
  it('renders without errors', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Tabs />, div);
    ReactDOM.unmountComponentAtNode(div);
  })
  //snapshot of tabs
  it('renders empty given no tabs', () => {
    const wrapper = shallow(<Tabs />)
    expect(toJson(wrapper)).toMatchSnapshot();
  })
  it('renders the first tab by default', ()=> {
    const wrapper = shallow(<Tabs tabs={TabsProp} />);
    expect(toJson(wrapper)).toMatchSnapshot()
  })
  it('closes the first tab and opens any clicked tab', ()=>{
    const wrapper = shallow(<Tabs tabs={TabsProp} />)
    wrapper.find('button');
    console.log('>>> WRAPPER <<<')
    console.log(wrapper.debug())
    console.log('>>> FIND(BUTTON) <<<')
    console.log(wrapper.find('button').at(1).simulate('click'))
    expect(toJson(wrapper)).toMatchSnapshot();
  })
})

/*
1. Made test file
2. Tabs component with empty div initially
3. import Tabs component to App && plugged in
prop, array of objects using dummy data
4. Rendering buttons in Tabs component
mapping over tabs array, making an array of buttons
using name of each object for each button's instance
5. set default props to fix Type Error: Cannot read 
property of undefined
6. Render first tab by default using static property
7. used discreet math to fix Error: Cannot read property of undefined
  using this.props.tabs.length && (Tabs.js line 18)
8. Create snapshot
  import renderer copying tabsProp from App.js to test file and snapshot
9. Added a click handler on the button
  updated snapshot by typing 'u' in terminal to fix error from having 1+ props on an element
10. Set intial set of buttons + added click event handler 
  (renderDOM is used between index.js and app.js)
*/