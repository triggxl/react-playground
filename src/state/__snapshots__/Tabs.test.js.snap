// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs Component closes the first tab and opens any clicked tab 1`] = `
<div>
  <button
    key="0"
    onClick={[Function]}
  >
    First tab
  </button>
  <button
    key="1"
    onClick={[Function]}
  >
    Second tab
  </button>
  <button
    key="2"
    onClick={[Function]}
  >
    Third tab
  </button>
  <div
    className="content"
  />
</div>
`;

exports[`Tabs Component renders empty given no tabs 1`] = `<div />`;

exports[`Tabs Component renders the first tab by dfault 1`] = `
<div>
  <button
    key="0"
    onClick={[Function]}
  >
    First tab
  </button>
  <button
    key="1"
    onClick={[Function]}
  >
    Second tab
  </button>
  <button
    key="2"
    onClick={[Function]}
  >
    Third tab
  </button>
  <div
    className="content"
  />
</div>
`;
